pkg/amphoracontrollers/daemonset.go

```diff
@@ -191,18 +193,32 @@ func DaemonSet(
 			},
 		},
 	}
-	// If possible two pods of the same service should not
-	// run on the same worker node. If this is not possible
-	// the get still created on the same worker node.
-	daemonset.Spec.Template.Spec.Affinity = affinity.DistributePods(
-		common.AppSelector,
-		[]string{
-			serviceName,
-		},
-		corev1.LabelHostname,
-	)
-	if instance.Spec.NodeSelector != nil {
-		daemonset.Spec.Template.Spec.NodeSelector = *instance.Spec.NodeSelector
+
+	if topology != nil {
+		// Get the Topology .Spec
+		ts := topology.Spec
+		// Process TopologySpreadConstraints if defined in the referenced Topology
+		if ts.TopologySpreadConstraints != nil {
+			daemonset.Spec.Template.Spec.TopologySpreadConstraints = *topology.Spec.TopologySpreadConstraints
+		}
+		// Process Affinity if defined in the referenced Topology
+		if ts.Affinity != nil {
+			daemonset.Spec.Template.Spec.Affinity = ts.Affinity
+		}
+	} else {
+		// If possible two pods of the same service should not
+		// run on the same worker node. If this is not possible
+		// the get still created on the same worker node.
+		daemonset.Spec.Template.Spec.Affinity = affinity.DistributePods(
```

Comment:
This doesn't make sense for a Daemonset. I guess we need to keep doing it to avoid changing the spec during updates. Maybe worth adding a comment to explain

Reply:
Right, I had a comment for myself earlier and I was looking for some resources I can link to the comment. I think it's fine  to state that even though we're keeping that code to ensure compatibility, in general `DaemonSets` follow different rules because they're different in nature from Deployments and StatefulSets.
In particular, `DaemonSets` automatically place one `Pod` on every node that matches the node selector, but topology spread constraints and affinity/anti-affinity rules are ignored.
I think DaemonSet pods bypass the default scheduler because of this [1].

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L1018

Reply:
resolving this one as I put an additional comment to explain we need to keep that code to be backward  compatible with what already deployed.